File: src/java/sequential/graph/dfs/IterativeDepthFirstTraversal.java
package sequential.graph.dfs;

import java.util.ArrayDeque;
import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;


public class IterativeDepthFirstTraversal {

    public static <T> Collection<T> dfs(Map<T, List<T>> graph, T root) {
        var explored = new LinkedHashSet<T>();
        var searchStack = new ArrayDeque<T>();

        searchStack.add(root);
        while (!searchStack.isEmpty()) {
            var node = searchStack.removeLast();
            if (!explored.contains(node)) {
                explored.add(node);
                searchStack.addAll(graph.getOrDefault(node, List.of()));
            }
        }
        return explored;
    }


    public static void main(String[] args) {
        var graph = Map.of(
                "you", List.of("alice", "bob", "clair"),
                "alice", List.of("peggy"),
                "clair", List.of("thom", "jonny"),
                "bob", List.of("anuj", "paggy")
        );

        System.out.println(
                dfs(graph, "you")
        );
    }
}

File: src/python/sequential/graph/dfs/iterative_depth_first_traversal.py
import collections


def dfs(graph, root):
    """
    Performs iterative depth-first traversal on the :param graph: starting from the :param root: node.
    """
    explored = set()
    traversed = []
    search_stack = collections.deque()

    search_stack.append(root)
    while search_stack:
        node = search_stack.pop()
        if node not in explored:
            explored.add(node)
            traversed.append(node)
            search_stack += graph.get(node, [])

    return traversed


def main():
    graph = {
        "you": ["alice", "bob", "clair"],
        "alice": ["peggy"],
        "clair": ["thom", "jonny"],
        "bob": ["anuj", "peggy"],
    }

    print(dfs(graph, "you"))

if __name__ == "__main__":
    main()
